src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,232 @@
+            .getTranslation()
+            .getNorm();
+    if (hasNote == true) {
+      if (this.inRange() == true) {
+        if (scoringPodium == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.PODIUM_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.PODIUM_VELOCITY);
+          io.setAngle(ShooterConstants.PODIUM_ANGLE);
+        } else if (scoringSubwoofer == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.SUBWOOFER_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.SUBWOOFER_VELOCITY);
+          io.setAngle(ShooterConstants.SUBWOOFER_ANGLE);
+        } else if (scoringAmp == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.AMP_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.AMP_VELOCITY);
+          io.setAngle(ShooterConstants.AMP_ANGLE);
+        } else if (aiming == true) {
+          io.setShooterWheelBottomVelocity(
+              veloTreeMap.get(RobotOdometry.getInstance().getEstimatedPosition().getX()));
+          io.setShooterWheelTopVelocity(
+              veloTreeMap.get(RobotOdometry.getInstance().getEstimatedPosition().getX()));
+          io.setAngle(angleTreeMap.get(RobotOdometry.getInstance().getEstimatedPosition().getX()));
+        } else { // not aiming
+          io.setShooterWheelBottomVelocity(ShooterConstants.SHOOTER_IDLE_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.SHOOTER_IDLE_VELOCITY);
+          io.setAngle(angleTreeMap.get(test));
+        }
+      } else { // not in range
+        io.setShooterWheelBottomVelocity(ShooterConstants.SHOOTER_IDLE_VELOCITY);
+        io.setShooterWheelTopVelocity(ShooterConstants.SHOOTER_IDLE_VELOCITY);
+        io.setAngle(ShooterConstants.SHOOTER_STORAGE_ANGLE);
+      }
+    } else { // no note
+      io.setAngle(ShooterConstants.SHOOTER_STORAGE_ANGLE);
+    }
+    this.goToConstantVelocity();
+  }
+
+  public void shoot(double topWheelVelocityRPS, double bottomWheelVelocityRPS) {
+    // needs to utilize kicker
+    io.setShooterWheelTopVelocity(topWheelVelocityRPS);
+    io.setShooterWheelBottomVelocity(bottomWheelVelocityRPS);
+  }
+
+  public void setAngle(double angle) {
+    io.setAngle(angle);
+  }
+
+  public void goToConstantVelocity() {
+    io.setShooterWheelTopVelocity(topWheelVelocity.get());
+    io.setShooterWheelBottomVelocity(bottomWheelVelocity.get());
+  }
+
+  public BooleanSupplier getShooterAngleReadySupplier() {
+    return () -> this.isAngleAtSetpoint();
+  }
+
+  public boolean readyToShoot() {
+    if (autoShooter == false) {
+      if (isTopShootAtSetpoint() && isBottomShootAtSetpoint() && isAimed) {
+        return true;
+      }
+      return false;
+    } else {
+      if (isTopShootAtSetpoint() && isBottomShootAtSetpoint() && isAngleAtSetpoint() && isAimed) {
+        return true;
+      }
+      return false;
+    }
+  }
+
+  public boolean inRange() {
```

Comment:
check the straight line distance to the speaker; use the code around line 90 so you don't need to know the alliance color